packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -568,13 +592,77 @@ Hardhat Network's forking functionality only works with blocks from at least spu
   ): Promise<EstimateGasResult> {
     // We get the CallParams and transform it into a TransactionParams to be
     // able to run it.
-    const txParams = {
-      ...callParams,
-      nonce: await this._getNonce(
-        new Address(callParams.from),
-        blockNumberOrPending
-      ),
-    };
+    const nonce = await this._getNonce(
+      new Address(callParams.from),
+      blockNumberOrPending
+    );
+
+    // TODO: This is more complex in Geth, we should make sure we aren't missing
+    //  anything here.
+
+    let gasPrice = callParams.gasPrice;
+    let maxFeePerGas = callParams.maxFeePerGas;
+    let maxPriorityFeePerGas = callParams.maxPriorityFeePerGas;
+
+    if (!this.isEip1559Active()) {
+      if (gasPrice === undefined) {
+        gasPrice = await this.getGasPrice();
+      }
+    } else if (gasPrice === undefined) {
+      if (maxPriorityFeePerGas === undefined) {
+        maxPriorityFeePerGas = await this.getMaxPriorityFeePerGas();
+
+        if (
+          callParams.maxFeePerGas !== undefined &&
+          callParams.maxFeePerGas.lt(maxPriorityFeePerGas)
+        ) {
+          maxPriorityFeePerGas = callParams.maxFeePerGas;
+        }
+      }
+
+      if (callParams.maxPriorityFeePerGas === undefined) {
```

Comment:
Does this `if` makes sense? It seems to me that, if if entered the previous one, then it will enter in this one, so it can just be removed and the two code blocks merged.

Reply:
Maybe this was meant to be `if (maxFeePerGas === undefined)`? If that's the case, it would mean that we need to add a test.

Reply:
it should be `maxFeePerGas`.